pkg/leader/leader.go

```diff
@@ -0,0 +1,191 @@
+// leader. Upon termination of that pod, the garbage collector will delete the
+// ConfigMap, enabling a different pod to become the leader.
+func Become(ctx context.Context, name string) error {
+	logrus.Info("trying to become the leader")
+
+	ns, err := myNS()
+	if err != nil {
+		return err
+	}
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return err
+	}
+
+	client, err := crclient.New(config, crclient.Options{})
+	if err != nil {
+		return err
+	}
+
+	owner, err := myOwnerRef(ctx, client, ns)
+	if err != nil {
+		return err
+	}
+
+	cm := &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            name,
+			Namespace:       ns,
+			OwnerReferences: []metav1.OwnerReference{owner},
+		},
+	}
+
+	// check for existing lock from this pod, in case we got restarted
+	existing := &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+	}
+	key := crclient.ObjectKey{Namespace: ns, Name: name}
+	err = client.Get(ctx, key, existing)
+
+	switch {
+	case err == nil:
+		for _, existingOwner := range existing.GetOwnerReferences() {
+			if existingOwner.Name == owner.Name {
+				logrus.Info("Found existing lock with my name. I was likely restarted.")
+				logrus.Info("Continuing as the leader.")
+				return nil
+			} else {
+				logrus.Infof("Found existing lock from %s", existingOwner.Name)
+			}
+		}
+	case apierrors.IsNotFound(err):
+		logrus.Info("No pre-existing lock was found.")
+	default:
+		logrus.Error("unknown error trying to get ConfigMap")
+		return err
+	}
+
+	// try to create a lock
+	seconds := 1
+	for {
+		err := client.Create(ctx, cm)
+		switch {
+		case err == nil:
+			logrus.Info("Became the leader.")
+			return nil
+		case apierrors.IsAlreadyExists(err):
+			logrus.Info("Not the leader. Waiting.")
+			select {
+			case <-time.After(wait.Jitter(time.Second*time.Duration(seconds), .2)):
+				if seconds < 16 {
+					seconds = seconds * 2
+				}
+				continue
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		default:
+			logrus.Error("unknown error creating configmap")
+			return err
+		}
+	}
+}
+
+// myNS returns the name of the namespace in which this code is currently running.
+func myNS() (string, error) {
+	nsBytes, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+	if err != nil {
+		if os.IsNotExist(err) {
+			return "", ErrNoNS
+		}
+		return "", err
+	}
+	ns := strings.TrimSpace(string(nsBytes))
+	logrus.Infof("found namespace: %s", ns)
+	return ns, nil
+}
+
+// myOwnerRef returns an OwnerReference that corresponds to the pod in which
+// this code is currently running.
+func myOwnerRef(ctx context.Context, client crclient.Client, ns string) (metav1.OwnerReference, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return metav1.OwnerReference{}, err
```

Comment:
`cannot use nil as type "github.com/operator-framework/operator-sdk/vendor/k8s.io/apimachinery/pkg/apis/meta/v1".OwnerReference in return argument`

Reply:
I see, the return arg is a type of `metav1.OwnerReference` not `*metav1.OwnerReference`? Maybe change to `*metav1.OwnerReference` as the return type?

Reply:
I think this is a normal pattern either way, and there are examples of returning an error with both nil-pointer and zero-value in the standard library. I'm not sure if you had any reasoning in mind beyond aesthetics? In any case, since metav1 itself only ever returns a pointer to an OwnerReference, I'll go ahead and change this to match.

Reply:
@mhrivnak from my past experience with Go, almost all the code I have seen has `return nil, err` pattern which is what we use in the operator-sdk case. i guess this is just a style nit to be consistent with rest of the codebase. Sorry I should have been more clear.